### PR TITLE
Fn & worker level concurrency

### PR DESCRIPTION
**Pull request by tonyhb**
## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

---

> [!NOTE]
> This PR adds worker-level concurrency control via app-scoped semaphores. Semaphore capacity is dynamically adjusted when workers connect/disconnect through a new `SemaphoreLifecycleListener`. Auto-release semaphores (worker concurrency, per-step) are propagated to all queue item types in the executor, while manual-release semaphores (function concurrency, per-run) are released at finalization. The Lua scripts are hardened with explicit weight validation and a capacity floor guard. The previous batching bug (different UsageValues in one Acquire call) is fixed by moving semaphore checks entirely out of `BacklogRefillConstraintCheck` and into per-item `ItemLeaseConstraintCheck`.
> 
> <sup>Written by [Mendral](https://mendral.com) for commit f8f64844f4e52a2c6b275881115e9a03367e4f7e.</sup>

### CODE REVIEW

File: pkg/execution/queue/constraints.go [hunk 1]
```diff
@@ -238,15 +231,42 @@ func (q *queueProcessor) BacklogRefillConstraintCheck(
 		}, nil
 	}
 
+	// Build constraint items and config. Semaphores come directly from queue items.
+	constraintItems := constraintItemsFromBacklog(shadowPart, backlog, constraints)
+	config := ConstraintConfigFromConstraints(constraints)
+
+	// Deduplicate semaphores across items — multiple start jobs for the same function
+	// would otherwise INCRBY the same semaphore counter multiple times in the Lua script.
```
Comment (BrunoScheufler) [inline pkg/execution/queue/constraints.go]: Do we test this? Feels like something gnarly 😄

Comment (tonyhb) [reply pkg/execution/queue/constraints.go]: we only allow 1 semaphore rn so this is literally an edge case from AI code review
